File: packages/hydra-python/src/main/python/hydra/dsl/meta/lib/strings.py
"""Phantom-typed term DSL for the hydra.lib.strings library."""

from hydra.phantoms import TTerm
from hydra.dsl.meta.phantoms import lib_primitive1 as primitive1, lib_primitive2 as primitive2


def cat(strings: TTerm) -> TTerm:
    """Concatenate a list of strings."""
    return primitive1(strings)


def cat2(s1: TTerm, s2: TTerm) -> TTerm:
    """Concatenate two strings."""
    return primitive2(s1, s2)


def char_at(index: TTerm, s: TTerm) -> TTerm:
    """Get the character at a given index in a string."""
    return primitive2(index, s)


def from_list(chars: TTerm) -> TTerm:
    """Convert a list of character codes to a string."""
    return primitive1(chars)


def intercalate(separator: TTerm, strings: TTerm) -> TTerm:
    """Insert a separator between strings and concatenate."""
    return primitive2(separator, strings)


def length(s: TTerm) -> TTerm:
    """Get the length of a string."""
    return primitive1(s)


def lines(s: TTerm) -> TTerm:
    """Split a string into lines."""
    return primitive1(s)


def null(s: TTerm) -> TTerm:
    """Check if a string is empty."""
    return primitive1(s)


def split_on(separator: TTerm, s: TTerm) -> TTerm:
    """Split a string on a separator."""
    return primitive2(separator, s)


def to_list(s: TTerm) -> TTerm:
    """Convert a string to a list of character codes."""
    return primitive1(s)


def to_lower(s: TTerm) -> TTerm:
    """Convert a string to lowercase."""
    return primitive1(s)


def to_upper(s: TTerm) -> TTerm:
    """Convert a string to uppercase."""
    return primitive1(s)


def unlines(strings: TTerm) -> TTerm:
    """Join lines with newline characters."""
    return primitive1(strings)

File: packages/hydra-java/src/main/java/hydra/dsl/meta/lib/Strings.java
package hydra.dsl.meta.lib;
import hydra.dsl.meta.Phantoms;
import hydra.typed.TypedTerm;

/** Phantom-typed term DSL for the {@code hydra.lib.strings} library. */
public final class Strings {
    private Strings() {}

    private static <A> TypedTerm<A> prim(String name) {
        return Phantoms.var("hydra.lib.strings." + name);
    }

    public static <A> TypedTerm<A> cat(TypedTerm<?> strings) { return Phantoms.apply(prim("cat"), strings); }
    public static <A> TypedTerm<A> cat2(TypedTerm<?> s1, TypedTerm<?> s2) { return Phantoms.apply(prim("cat2"), s1, s2); }
    public static <A> TypedTerm<A> fromList(TypedTerm<?> chars) { return Phantoms.apply(prim("fromList"), chars); }
    public static <A> TypedTerm<A> intercalate(TypedTerm<?> separator, TypedTerm<?> strings) { return Phantoms.apply(prim("intercalate"), separator, strings); }
    public static <A> TypedTerm<A> length(TypedTerm<?> s) { return Phantoms.apply(prim("length"), s); }
    public static <A> TypedTerm<A> lines(TypedTerm<?> s) { return Phantoms.apply(prim("lines"), s); }
    public static <A> TypedTerm<A> maybeCharAt(TypedTerm<?> index, TypedTerm<?> s) { return Phantoms.apply(prim("maybeCharAt"), index, s); }
    public static <A> TypedTerm<A> null_(TypedTerm<?> s) { return Phantoms.apply(prim("null"), s); }
    public static <A> TypedTerm<A> splitOn(TypedTerm<?> separator, TypedTerm<?> s) { return Phantoms.apply(prim("splitOn"), separator, s); }
    public static <A> TypedTerm<A> toList(TypedTerm<?> s) { return Phantoms.apply(prim("toList"), s); }
    public static <A> TypedTerm<A> toLower(TypedTerm<?> s) { return Phantoms.apply(prim("toLower"), s); }
    public static <A> TypedTerm<A> toUpper(TypedTerm<?> s) { return Phantoms.apply(prim("toUpper"), s); }
    public static <A> TypedTerm<A> unlines(TypedTerm<?> strings) { return Phantoms.apply(prim("unlines"), strings); }
}
